### PR TITLE
Providers resource

### PR DESCRIPTION
**Pull request by enricostano**
Before placing an `Order` to a given `Producer` the users of a `Group` need to retrieve first all the `Producer`s that are available for a `Group` as providers:
```
GET /api/v1/producers?group_id=27
```
This request will return all the `Producer`s associated to a `Group` through the join model `Supplier`.
```json
[
  {
    "id":12,
    "name":"Producer #1",
    "address":"C/ del primer productor",
    "email":"productor+1@katuma.org",
    "created_at":"2016-09-19T18:19:46.233Z",
    "updated_at":"2016-09-19T18:19:46.233Z"
  },
  {
    "id":15,
    "name":"Producer #2",
    "address":"C/ del segundo productor",
    "email":"productor+2@katuma.org",
    "created_at":"2016-09-19T18:19:46.233Z",
    "updated_at":"2016-09-19T18:19:46.233Z"
  }
]
```

### CODE REVIEW

**Comment by andresgutgon** [comment]
:+1:

**Comment by sauloperez** [comment]
👍

File: engines/suppliers/app/controllers/suppliers/api/v1/providers_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+          authorize @group
+        end
+
+        def load_producer
+          @producer = ::Suppliers::Producer.find_by_id(params[:id])
+
+          return head :not_found unless @producer
+
+          authorize @producer
```
Comment (andresgutgon) [inline engines/suppliers/app/controllers/suppliers/api/v1/providers_controller.rb]: Why not get normal Producer `/API/producers/:id` instead of Provider::Producer?

Comment (enricostano) [inline engines/suppliers/app/controllers/suppliers/api/v1/providers_controller.rb]: I don't understand the question, sorry :pray: 

Here `@producer` is an instance of `Suppliers::Producer` model, that points directly to the `producers` table.

Comment (enricostano) [inline engines/suppliers/app/controllers/suppliers/api/v1/providers_controller.rb]: Furthermore, since we're in the `Suppliers::ProvidersController`, all the operations that we'll do with `@producer` will be scoped under the `Suppliers` namespace since we're in that engine. In this way if the concept of `Producer` needs some specific business logic, only related to this `Suppliers` context we'll be able to keep it clean qithout polluting the `Producers::Producer` model.

Even the `current_user` here is an instance of `Suppliers::User` and not of `Account::User` (which is the actual *owner* of the `users` table).

Better? Useful?

Comment (andresgutgon) [inline engines/suppliers/app/controllers/suppliers/api/v1/providers_controller.rb]: Is clear but...

> In this way if the concept of Producer needs some specific business logic, only related to this Suppliers context, we'll be able to keep it clean qithout polluting the Producers::Producer model.

Right now do we have some specific logic in this `show`? If the answer is no why do an extra end point `/show` if we already have original `api/producers/:id`.

Is just a question, I'm not saying to change anything 😄

File: engines/suppliers/spec/controllers/suppliers/api/v1/providers_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            it_behaves_like 'a successful request'
+
+            describe 'its body' do
+              before { get :show, id: producer.id }
+
+              subject { JSON.parse(response.body) }
+
+              it { is_expected.to eq(JSON.parse(producer.to_json)) }
+            end
+          end
```
Comment (sauloperez) [inline engines/suppliers/spec/controllers/suppliers/api/v1/providers_controller_spec.rb]: Wouldn't it be easier to follow if we do all the examples for `show` and `index` in their own describe? As it is you need to keep switching context between these actions and makes it very hard. Besides, it is not consistent with the other tests, which makes it even harder for new devs.

File: engines/suppliers/app/controllers/suppliers/api/v1/providers_controller.rb (hunk 1 above)
Comment (enricostano) [inline engines/suppliers/app/controllers/suppliers/api/v1/providers_controller.rb]: In this specific case the `show` and `index` endpoints in `providers` controller use the `suppliers` relationship. The `producers` controller `show` and `index` use the `memberships` relation instead.

The first relation, `suppliers`, tells you if the user can use that `producer` as a provider and place orders.
The second one, `memberships`, tells you if the user ha admin authority on the given `producer`. Useful for instance to change the name of the producer or add a product associated to it.

The data model is a bit complex here since `producer` model has two kind of "join tables". And we need that since one (`producers_memberhips`) represents the ownership of the `producer` (with `users` or `groups`), while the other (`suppliers`) represents the relationships as a provider between a `producer` ad a `group`.

File: engines/suppliers/spec/controllers/suppliers/api/v1/providers_controller_spec.rb (hunk 2 above)
Comment (enricostano) [inline engines/suppliers/spec/controllers/suppliers/api/v1/providers_controller_spec.rb]: They are already in two separated describe:
```ruby
 59   describe 'GET #index' do
  ...
 80   end
  ...
 82   describe 'GET #show' do
  ...
101   end
```
Isn't that enough?

Comment (sauloperez) [inline engines/suppliers/spec/controllers/suppliers/api/v1/providers_controller_spec.rb]: What I found a bit confusing was having the un/authenticated contexts and then the show and index inside. I though there were more describe blocks so, nevermind.

File: engines/producers/app/models/producers/product.rb [hunk 3]
```diff
@@ -2,15 +2,17 @@ module Producers
...
-    }
+    }.freeze
+
+    belongs_to :producer, class_name: 'Producers::Producer'.freeze
 
     # TODO: move price to its own table/s
-    validates :name, :price, :unit, presence: true
+    validates :name, :price, :unit, :producer, presence: true
+    validates :unit, inclusion: { in: UNITS.values }
+    validates_numericality_of :price, :unit
```
Comment (sauloperez) [inline engines/producers/app/models/producers/product.rb]: Isn't there the equivalent validation with the newer `validates` syntax?

File: engines/producers/app/policies/producers/product_policy.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      user_admin_membership_for?(user) || group_admin_membership_for?(user)
+    end
+
+    def destroy?
+      user_admin_membership_for?(user) || group_admin_membership_for?(user)
+    end
+
+    private
+
+    def user_membership_for?(user)
```
Comment (sauloperez) [inline engines/producers/app/policies/producers/product_policy.rb]: Looks like something that we can DRY together with the producer policy. Btw, does pundit impose any limitation on defining the initialize method in the concrete policy classes? I assume not

File: engines/producers/app/services/producers/product_finder.rb [hunk 5]
```diff
@@ -6,7 +6,7 @@ class ProductFinder
     # @param producer [Producers::Producer]
     # @return [ActiveRecord::Relation<Producers::Product>]
     def find_by_producer(producer)
-      Product.where(producer: producer)
+      Product.where(producer_id: producer.id)
```
Comment (sauloperez) [inline engines/producers/app/services/producers/product_finder.rb]: I believe it's indifferent whether you use the id or not. What's the reason for the change?

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+          before { authenticate_as producers_user }
+
+          describe 'GET #index' do
+            let!(:first_product) do
+              FactoryGirl.create(
+                :product,
+                producer: producer_for_group
+              )
+            end
+            let!(:second_product) do
```
Comment (sauloperez) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: Please, try to give more meaningful names :pray:

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+
+            it_behaves_like 'a successful request'
+
+            describe 'its body' do
+              before { get :index, producer_id: producer_for_group.id }
+
+              subject { JSON.parse(response.body) }
+
+              its(:size) { is_expected.to eq(2) }
+              it { is_expected.to include(JSON.parse(first_product.to_json)) }
```
Comment (sauloperez) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: Why are we converting to json and then parsing it back? *Menos por menos más*.

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+
+          describe 'POST #create' do
+            subject { post :create, params }
+
+            context 'when the producer is directly associated to the user' do
+              let(:params) do
+                {
+                  name: 'Tomatekken',
+                  price: 10.13,
+                  unit: Product::UNITS[:kg],
```
Comment (sauloperez) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: IMO this :nose: a lot. In the models file already looked like a responsibility to extract, but know turns out quite evident to me. 

Here we're reaching into inner state of product and coupling to the data structure of the units. This makes it hard to change it, and by the time you realize, it won't be possible to refactor as it'll bé used everywhere.

So, we should encapsulate the data structure under a simple api

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+                its(['producer_id']) { is_expected.to eq(params[:producer_id]) }
+              end
+
+              context 'and the user is not a producer admin' do
+                before do
+                  producer_membership = Membership.where(
+                    user: producers_user,
+                    producer: producer_for_user
+                  ).first
+                  producer_membership.role = Membership::ROLES[:member]
```
Comment (sauloperez) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: It looks the same as units

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+              end
+
+              it_behaves_like 'a bad request'
+
+              describe 'its body' do
+                before { post :create, params }
+
+                subject { JSON.parse(response.body) }
+
+                its(:size) { is_expected.to eq(2) }
```
Comment (sauloperez) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: Imo we can get rid of this expectation by simply checking that the response contains exactly the two elements below. You can use at least match_array or contaib_exactly matchers. This will save precious performance

File: Gemfile [hunk 11]
```diff
@@ -17,7 +17,7 @@ gem 'sidekiq', '4.0.1'
 
 gemspec path: 'engines/shared'
 gemspec path: 'engines/account'
-gemspec path: 'engines/group'
+gemspec path: 'engines/basic_resources'
```
Comment (andresgutgon) [inline Gemfile]: ❤️

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb (hunk 8 above)
Comment (enricostano) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: I agree, but just don't stick to the Rails way since we're exploring something else here.

File: engines/producers/app/policies/producers/product_policy.rb (hunk 4 above)
Comment (enricostano) [inline engines/producers/app/policies/producers/product_policy.rb]: We could even move it to the model in the specific engine, like `producer.is_admin?(user)`

File: engines/producers/app/services/producers/product_finder.rb (hunk 5 above)
Comment (enricostano) [inline engines/producers/app/services/producers/product_finder.rb]: just `producer` doesn't work, will investigate why... but not now

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb (hunk 8 above)
Comment (enricostano) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: We do this to assign a meaningful thing instead of a plain `0`

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb (hunk 9 above)
Comment (enricostano) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: We do this to assign a meaningful thing instead of a plain `0`

File: engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb (hunk 6 above)
Comment (enricostano) [inline engines/producers/spec/controllers/producers/api/v1/products_controller_spec.rb]: Ehm... *manzana*?

File: engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        #
+        # Without parameters it will return all the producers which the
+        # user is associated to directly as a member.
+        #
+        # Passing `group_id` parameter will return:
+        #  - the producers associated to the group through `Membership`
+        #  - the producers associated to the group through `Supplier`
+        #
+        def index
+          producers = ProducersCollection.new(user: current_user, group: @group).build
```
Comment (sauloperez) [inline engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb]: So this is the "black box" we were talking about that merges two results sets?

File: engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
+module Suppliers
+  module Api
+    module V1
+      class ProducersController < ApplicationController
+        before_action :authenticate
+        before_action :load_group,
+          only: [:index],
+          if: -> { !!params.fetch('group_id', nil) }
```
Comment (sauloperez) [inline engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb]: When there are so many exceptions, IMO there's no point on having it in a before_action and it's way better to explicitly call `load_group` from the index action; easier to read.

File: engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        # user is associated to directly as a member.
+        #
+        # Passing `group_id` parameter will return:
+        #  - the producers associated to the group through `Membership`
+        #  - the producers associated to the group through `Supplier`
+        #
+        def index
+          producers = ProducersCollection.new(user: current_user, group: @group).build
+
+          render json: ::Suppliers::ProducersSerializer.new(producers)
```
Comment (sauloperez) [inline engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb]: did you try without the `::Suppliers::`?

File: engines/suppliers/spec/controllers/suppliers/api/v1/producers_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+              subject { get :show, id: producer_for_group.id }
+
+              it_behaves_like 'a successful request'
+
+              describe 'its body' do
+                before { get :show, id: producer_for_group.id }
+
+                subject { JSON.parse(response.body) }
+
+                it { is_expected.to eq(JSON.parse(ProducerSerializer.new(producer_for_group).to_json)) }
```
Comment (sauloperez) [inline engines/suppliers/spec/controllers/suppliers/api/v1/producers_controller_spec.rb]: ⚠️  long line according to code climate

File: engines/suppliers/spec/controllers/suppliers/api/v1/producers_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+              subject { get :show, id: producer_for_user.id }
+
+              it_behaves_like 'a successful request'
+
+              describe 'its body' do
+                before { get :show, id: producer_for_user.id }
+
+                subject { JSON.parse(response.body) }
+
+                it { is_expected.to eq(JSON.parse(ProducerSerializer.new(producer_for_user).to_json)) }
```
Comment (sauloperez) [inline engines/suppliers/spec/controllers/suppliers/api/v1/producers_controller_spec.rb]: ⚠️  long line according to code climate

File: engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb (hunk 12 above)
Comment (enricostano) [inline engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb]: Yep, this is a first implementation:
 - `ProducersCollection` 's responsibility is to retrieve the producers based on the presence of `group`
 - `ProducerSerializer` will decorate the model with the needed flags

What do you think?

File: engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb (hunk 13 above)
Comment (enricostano) [inline engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb]: Yes, I agree :clap:

File: engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb (hunk 12 above)
Comment (sauloperez) [inline engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb]: Looks good. Although I saw that can surely implement the "merge" with a single query, which would be better in the long run, but we can wait.. a lot :trollface:

**Comment by andresgutgon** [comment]
CTO === Codecliamte?
image

File: engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb (hunk 12 above)
Comment (enricostano) [inline engines/suppliers/app/controllers/suppliers/api/v1/producers_controller.rb]: Not sure we can do all in one query since we touch 3 tables: `producers`, `suppliers` and `memberships`. But we can investigate... just not now :sweat_smile:

File: engines/suppliers/app/presenters/suppliers/producer_presenter.rb [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    # @param user [User]
+    def initialize(producer, user)
+      @producer = producer
+      @user = user
+    end
+
+    # Checks if the given user has edit rights on the producer
+    #
+    # @return [Boolean]
+    def can_edit
```
Comment (sauloperez) [inline engines/suppliers/app/presenters/suppliers/producer_presenter.rb]: add the `?`, please

File: engines/suppliers/app/presenters/suppliers/producer_presenter.rb [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+      producer.has_admin?(user) || group_admin?
```
Comment (sauloperez) [inline engines/suppliers/app/presenters/suppliers/producer_presenter.rb]: and the call to `#has_admin?` introduces an N+1, doesn't it?

File: engines/suppliers/app/presenters/suppliers/producer_presenter.rb [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    end
+
+    delegate :id, :name, :email, :address, :created_at, :updated_at, to: :producer
```
Comment (sauloperez) [inline engines/suppliers/app/presenters/suppliers/producer_presenter.rb]: We might need to use either SimpleDelegator overriding the constructor or use our own Delegator subclass. This doesn't follow open/closed, meaning that every time we add a new column in the producers table, we'll need to add it here as well.

File: engines/suppliers/app/presenters/suppliers/producer_presenter.rb [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    attr_reader :producer, :user
+
+    # Checks if the producer is associated to a group
+    # and if the user is an admin of that group
+    #
+    # @return [Boolean]
+    def group_admin?
+      return false unless group
+
+      group.has_admin?(user)
```
Comment (sauloperez) [inline engines/suppliers/app/presenters/suppliers/producer_presenter.rb]: `group && group.has_admin?(user)` right? and yet another N+1, do the presenters eager load some how?

File: engines/suppliers/app/presenters/suppliers/producer_presenter.rb (hunk 17 above)
Comment (enricostano) [inline engines/suppliers/app/presenters/suppliers/producer_presenter.rb]: it's an attribute, for now the serializers are just mapping attributes from models/presenters directly. Maybe an alias or something. will check.

File: engines/suppliers/app/presenters/suppliers/producer_presenter.rb (hunk 19 above)
Comment (enricostano) [inline engines/suppliers/app/presenters/suppliers/producer_presenter.rb]: Sure, this is just a proptotype. We may drop the presenter and flag thing all the way.

File: engines/suppliers/app/presenters/suppliers/producer_presenter.rb (hunk 20 above)
Comment (enricostano) [inline engines/suppliers/app/presenters/suppliers/producer_presenter.rb]: yep
